libs/metrics-wai/src/Data/Metrics/Middleware/Prometheus.hs

```diff
@@ -0,0 +1,34 @@
+module Data.Metrics.Middleware.Prometheus (waiPrometheusMiddleware) where
+
+import           Imports
+import qualified Network.Wai                       as Wai
+import           Network.Wai.Routing.Route         (Routes, prepare)
+import qualified Network.Wai.Middleware.Prometheus as Promth
+import qualified Data.Text.Encoding                as T
+
+import Data.Metrics.WaiRoute (treeToPaths)
+import Data.Metrics.Types (treeLookup)
+
+-- | Adds a prometheus metrics endpoint at @/i/metrics@
+-- This middleware requires your servers 'Routes' because it does some normalization
+-- (e.g. removing params from calls)
+waiPrometheusMiddleware :: Monad m => Routes a m b -> Wai.Middleware
+waiPrometheusMiddleware routes =
+    Promth.prometheus conf . Promth.instrumentHandlerValue (normalizeWaiRequestRoute routes)
+  where
+    conf = Promth.def
+        { Promth.prometheusEndPoint      = ["i", "metrics"]
+          -- We provide our own instrumentation so we can normalize routes
+        , Promth.prometheusInstrumentApp = False
+        }
+
+-- | Compute a normalized route for a given request. 
+-- Normalized routes have route parameters replaced with @<>@
+normalizeWaiRequestRoute :: Monad m => Routes a m b -> Wai.Request -> Text
+normalizeWaiRequestRoute routes req = pathInfo
+  where
+    mPathInfo :: Maybe ByteString
+    mPathInfo = treeLookup (treeToPaths $ prepare routes) (T.encodeUtf8 <$> Wai.pathInfo req)
+
+    pathInfo :: Text
+    pathInfo  = T.decodeUtf8 $ fromMaybe "N/A" mPathInfo
```

Comment:
This causes paths we haven't defined to show as `N/A`; which might not be so helpful for debugging.

Is there any downside for if I use the path from the request in the failure case? Or do we not want to pollute our metrics? Kinda seems like it would be helpful to still have SOME info about which routes are 404'ing.

Reply:
but aren't 404s reported with path?  `mPathInfo` is considered if Just, only if we don't have the information do we put in "N/A".

if i misunderstand and there we have path info we don't use, i agree with you and suggest we use it.

Reply:
@fisx perhaps they're reported in logs that way somehow, but I don't think so (what would be logging it?).

Here's the metrics output for a 404:

```
http_request_duration_seconds_bucket{handler="N/A",method="GET",status_code="404",le="10.0"} 1
```

I've updated it to the more helpful:

```
http_request_duration_seconds_bucket{handler="/my-broken-route",method="GET",status_code="404",le="10.0"} 1
```
This will still only fire for calls that make it through nginx; so it's likely to indicate a problem with a client if we get a lot of these.

Reply:
ok, seems right to me.  maybe we did the `N/A` in the past to have one figure for all stray requests, but i presume that can be aggregated in prometheus?

Reply:
I don't think we even necessarily want to aggregate them all, they may be independently interesting; we shouldn't be getting any 404's at all unless a client has gone silly; nginx should catch most of them.